src/main/java/eu/europeana/metis/sandbox/service/dataset/DefaultDatasetGenerator.java

```diff
@@ -0,0 +1,53 @@
+package eu.europeana.metis.sandbox.service.dataset;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import eu.europeana.metis.sandbox.common.Status;
+import eu.europeana.metis.sandbox.common.Step;
+import eu.europeana.metis.sandbox.common.locale.Country;
+import eu.europeana.metis.sandbox.common.locale.Language;
+import eu.europeana.metis.sandbox.domain.Dataset;
+import eu.europeana.metis.sandbox.domain.Record;
+import eu.europeana.metis.sandbox.service.util.XmlRecordProcessorService;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.springframework.stereotype.Service;
+
+@Service
+class DefaultDatasetGenerator implements DatasetGeneratorService {
+
+  private final XmlRecordProcessorService xmlRecordProcessorService;
+
+  public DefaultDatasetGenerator(
+      XmlRecordProcessorService xmlRecordProcessorService) {
+    this.xmlRecordProcessorService = xmlRecordProcessorService;
+  }
+
+  @Override
+  public Dataset generate(String name, Country country, Language language,
+      List<String> records) {
+    requireNonNull(name, "Dataset name must not be null");
+    requireNonNull(country, "Country must not be null");
+    requireNonNull(language, "Language must not be null");
+    requireNonNull(records, "Records must not be null");
+    checkArgument(!records.isEmpty(), "Records must not be empty");
+
+    var datasetId = System.currentTimeMillis() + "_" + name;
```

Comment:
I don't think we should have the name be part of the dataset ID, as it is not like this in Metis either and I think it may cause issues as we are not expecting a '_' in the dataset ID. I think we should stick to a number. 

Perhaps there is a way to not have to use the system time either, given that theoretically we could have collisions if two people send a dataset at the same time. This is, however, extremely unlikely. 

Can't we let the database figure this out (i.e. autoincrement)?

Reply:
Sorry, I guess I did not understand the dataset Id structure. so dataset Id is just a number ? at the beginning we agreed that id was going to be mix of generated number and name so that we retrieve that to the client to perform monitoring, that I why I was not concerned with collisions (I mean not super concerned because then 2 clients will have to use the same name also). And yes we can let the DB take care of it, but I will go to you to clarify some knowledge

Reply:
Well actually the facts are:
- DatasetId in metis is a String value, that is a number with a potential suffix of a letter. We generally manipulate them as unique numbers.
- The edm_datasetName which is part of the xml of the record is a concatenated value of the datasetId and the datasetName with an underscore in between. That field is what is used when searching records for a specific dataset in the europeana portal. It is not ideal of course but this is how it's implemented at the moment.

Hope that clears out the situation a bit.

Reply:
Noting here, but as a general note, why use "Default" prefix on the classes? I would avoid using that prefix and if there is another class that extends a base class then that extended class should have a descriptive name.

Reply:
For those that is the default implementation of the interface, in this case the interface is called DatasetGenerator and this class is called DefaultDatasetGenerator, while there is debate on [naming conventions](https://stackoverflow.com/questions/2814805/java-interfaces-implementation-naming-convention) this is the one I go for, mostly to go for [Dependency Inversion](https://en.wikipedia.org/wiki/Dependency_inversion_principle) principle, so the trade-off is that I name implementations DefaultInterfaceName just because right now that is the only implementation available, if there are more than one implementation in the future then name could evolve. And last while some people use the suffix Impl or name their interfaces with prefix I, I never liked those because at the end the interface is the thing you want to represent, like an example a guy said about List interface. That is my reasoning behind it, but we can talk about it if you want

Reply:
As the link you shared says about suffixing(or prefixing in your case):

> If all you can come up with to make your Class name unique is suffixing it with Impl, then you need to rethink having an Interface at all. So when you have a situation where you have an Interface and a single Implementation that is not uniquely specialized from the Interface you probably don't need the Interface.

Not saying in metis we are doing it absolutely right.. We try to avoid those prefixes/suffixes or general interfaces if they are not required( in general or not yet).
That being said, we actually do have some classes that have a suffix of "Impl", so as mentioned already we are far from being "perfect". :D
Personally if I had to choose between "Default" prefix and "Impl" suffix I would got for the latter just because the interface and implementation  are easier directory listed and searched because they are very close to each other since they start with the same name..

So yeah in short, this was a observation and we are flexible in general but since we do not have "Default" prefix in metis, if you want to keep the interface then I would suggest to have an "Impl" suffix instead, just for consistency..

Reply:
I think this is a something everyone faces, that is why there is debate and people has different opinions on it. Some people use Default, others Impl, others avoid using the interfaces, and there is no general consent about a best practice (at least from my point of view), but I prefer to preserve the interfaces so that leans more into [SOLID](https://en.wikipedia.org/wiki/SOLID) which drives into more testable code, again for this topic are different opinions, but `code to interfaces not to implementations` is a practice that also interacts good with Sprint DI.
Saying that I will switch implementations Impl suffix to follow the convention.
Thanks for sharing 👍